lualib/resty/events/broker.lua

```diff
@@ -146,40 +171,98 @@ local _MT = { __index = _M, }
 function _M.new(opts)
     return setmetatable({
         _opts = opts,
+        _queues = nil,
         _uniques = nil,
         _clients = nil,
+        _first_worker_id = nil,
+        _last_worker_id = nil,
     }, _MT)
 end
 
 function _M:init()
-    assert(self._opts)
+    local opts = self._opts
+
+    assert(opts)
 
     local _uniques, err = lrucache.new(MAX_UNIQUE_EVENTS)
     if not _uniques then
         return nil, "failed to create the events cache: " .. (err or "unknown")
     end
 
+    local queues = {}
+
+    local first_worker_id = opts.enable_privileged_agent == true and -1 or 0
+    local last_worker_id = worker_count() - 1
+
+    for i = first_worker_id, last_worker_id do
+        queues[i] = queue.new(opts.max_queue_len)
+    end
+
     self._uniques = _uniques
     self._clients = setmetatable({}, WEAK_KEYS_MT)
+    self._queues = queues
+    self._first_worker_id = first_worker_id
+    self._last_worker_id = last_worker_id
+
+    log(NOTICE, "event broker is ready to accept connections on worker #", opts.broker_id)
 
     return true
 end
 
 function _M:run()
+    local broker_id = ngx_worker_id()
+    if broker_id ~= self._opts.broker_id then
+        log(ERR, "broker got connection from worker on non-broker worker #", broker_id)
+        return exit(444)
+    end
+
+    local clients = self._clients
+    if not clients then
+        log(ERR, "broker is not (yet) ready to accept connections on worker #", broker_id)
+        return exit(444)
+    end
+
     local worker_connection, err = server.new()
     if not worker_connection then
         log(ERR, "failed to init socket: ", err)
-        exit(444)
+        return exit(444)
     end
 
-    self._clients[worker_connection] = queue.new(self._opts.max_queue_len)
+    local queues = self._queues
+    local worker_id = worker_connection.info.id
+    local worker_pid = worker_connection.info.pid
+
+    if worker_id == -1 and not queues[-1] then
+        -- TODO: this is for backward compatibility
+        --
+        -- Queue for the privileged agent is dynamically
+        -- created because it is not always enabled or
+        -- does not always connect to broker. This also
+        -- means that privileged agent may miss some
+        -- events on a startup.
+        --
+        -- It is suggested to instead explicitly pass
+        -- an option: enable_privileged_agent=true|false.
+        queues[-1] = queue.new(self._opts.max_queue_len)
+        self._first_worker_id = -1
+    end
+
+    clients[worker_connection] = true
 
     local read_thread_co = spawn(read_thread, self, worker_connection)
-    local write_thread_co = spawn(write_thread, self, worker_connection)
+    local write_thread_co = spawn(write_thread, self, worker_connection, queues[worker_id])
+
+    if worker_id == -1 then
+        log(NOTICE, "privileged agent connected to events broker (worker pid: ",
```

Comment:
Should we log the worker ID instead of PID? Or, in the log lines above, use the PID instead of worker ID? I can see the appeal of using the PID since that is already logged in the nginx logs, but the worker ID is not.